p4-16/spec/P4-16-spec.mdk

```diff
@@ -6416,6 +6416,153 @@ ck16.update( { h.ipv4.ttl, h.ipv4.proto } );
 h.ipv4.hdrChecksum = ck16.get();
 ~ End P4Example
 
+
+# Appendix: Restrictions on compile time and run time calls
+
+This appendix summarizes restrictions on compile time and run time
+calls that can be made.  Many of them are described earlier in this
+document, but are collected here for easy reference.
+
+The stateful types of objects in P4~16~ are packages, parsers,
+controls, externs, and tables.  All other types are referred to as
+"value types" here.
+
+Some guiding principles:
+
+- Controls are not allowed to call parsers, and vice versa, so there
+  is no use in passing one type to the other in constructor parameters
+  or run-time parameters.
+- At run time, after a control is called, and before that call is
+  complete, there can be no recursive calls between controls, nor from
+  a control to itself.  Similarly for parsers.  There can be loops
+  among states within a single parser.
+- Externs are not allowed to call parsers or controls, so there is no
+  use in passing objects of those types to them.
+- Tables are always instantiated directly in their enclosing control,
+  and cannot be instantiated at the top level.  There is no syntax for
+  specifying parameters that are tables.  Tables are only intended to
+  be used from within the control where they are defined.
+
+A note on recursion: It is expected that some architectures will
+define capabilities for recirculating a packet to be processed again
+as if it were a newly arriving packet, or to make "clones" of packets
+that are then processed by parsers and/or control blocks that the
+original packet has already completed.  This does not change the notes
+above on recursion that apply while a parser or control is executing.
+
+The first table lists restrictions on what types can be passed as
+constructor parameters to other types.
+
+|-------------|-------------------------------------------------|
+|             | can be a constructor parameter for this type ||||
+|             |---------|--------|---------|--------|
+| This type   | package | parser | control | extern |
++-------------+:-------:+:------:+:-------:+:------:+
+| package     | yes     |  no    |  no     |  no    |
+| parser      | yes     | yes    |  no     |  no    |
+| control     | yes     |  no    | yes     |  no    |
+| extern      | yes     | yes    | yes     | yes    |
+| table       |  no     |  no    |  no     |  no    |
+| value types | yes     | yes    | yes     | yes    |
+|-------------|---------|--------|---------|--------|
+
+The next table lists restrictions on where one may perform
+instantiations (see Section [#sec-instantiations]) of different types.
+The answer for `package` is always "no" because there is no
+"inside a package" where instantiations can be written in P4~16~.  One
```

Comment:
No package inside a package is a limitation that will haunt us. We briefly touched on how hard is to express architectures with multiple repeated components (e.g. multiple pipelines) and being able to express them as nested packages will greatly simplify the definition of such architectures. I realize that it is not currently used in PSA, but do we really want to fully restrict doing it in the language?

I'm thinking about the following example:
```
const N_PIPES = 8;
package Pipe (Parser p, Ingress i, Egress e, Deparser d);
package MultiPipe (package Pipes p[N_PIPES]);
```
I do read this as an instantiation within the package. It is true that it is unlikely that these will be instantiations by a user program, but rather they'll be instantiated by the architecture. However, for better or worse, we are using the _same_ language for both user programs and architecture definitions. Are we at the point where we should consider certain language features to be applicable only in architecture definition files?